File: DCC2024_Artur/January/Day11_Maximum_Difference_Between_Node_and_Ancestor.py
"""
https://leetcode.com/problems/maximum-difference-between-node-and-ancestor/description/?envType=daily-question&envId=2024-01-11
"""

from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def __init__(self) -> None:
        self.maxDiff = 0

    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:
        self.maxDiff = 0
        self.postorder(root)
        return self.maxDiff
        
    def postorder(self, node: TreeNode) -> tuple:
        currMin = node.val
        currMax = node.val
        
        if not node.left and not node.right:
            return (currMin, currMax)
        
        if node.left:
            left = self.postorder(node.left)
            self.maxDiff = max(self.maxDiff, abs(node.val - left[0]))
            self.maxDiff = max(self.maxDiff, abs((node.val - left[1])))
            currMin = min(currMin, left[0])
            currMax = max(currMax, left[1])
            
        if node.right:
            right = self.postorder(node.right)
            self.maxDiff = max(self.maxDiff, abs(node.val - right[0]))
            self.maxDiff = max(self.maxDiff, abs(node.val - right[1]))
            currMin = min(currMin, right[0])
            currMax = max(currMax, right[1])
            
        return (currMin, currMax)
        
        
if (__name__ == '__main__'):
    solution = Solution()

    tree1Node4 = TreeNode(4)
    tree1Node7 = TreeNode(7)
    tree1Node1 = TreeNode(1)
    tree1Node13 = TreeNode(13)
    tree1Node6 = TreeNode(6, tree1Node4, tree1Node7)
    tree1Node14 = TreeNode(14, tree1Node13, None)
    tree1Node3 = TreeNode(3, tree1Node1, tree1Node6)
    tree1Node10 = TreeNode(10, None, tree1Node14)
    tree1Node8 = TreeNode(8, tree1Node3, tree1Node10)
    
    tree2Node3 = TreeNode(3)
    tree2Node0 = TreeNode(0, tree2Node3, None)
    tree2Node2 = TreeNode(2, None, tree2Node0)
    tree2Node1 = TreeNode(1, None, tree2Node2)
    
    print(solution.maxAncestorDiff(tree1Node8))
    print(solution.maxAncestorDiff(tree2Node1))

File: DCC2024_Artur/January/Day11_Maximum_Difference_Between_Node_and_Ancestor.cpp
/*
 https://leetcode.com/problems/maximum-difference-between-node-and-ancestor/description/?envType=daily-question&envId=2024-01-11
*/

#include <iostream>
#include <vector>

using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};


class Solution
{
private:
    int maxDiff;

public:
    int maxAncestorDiff(TreeNode* root)
    {
        maxDiff = 0;
        preorder(root);

        return maxDiff;
    }

    pair<int, int> preorder(TreeNode* root)
    {
        int currMin = root->val;
        int currMax = root->val;

        if( !root->left && !root->right )
            return make_pair(currMin, currMax);

        if( root->left )
        {
            auto left = preorder(root->left);
            maxDiff = max(maxDiff, abs(root->val - left.first));
            maxDiff = max(maxDiff, abs(root->val - left.second));
            currMin = min(currMin, left.first);
            currMax = max(currMax, left.second);
        }
        if( root->right )
        {
            auto right = preorder(root->right);
            maxDiff = max(maxDiff, abs(root->val - right.first));
            maxDiff = max(maxDiff, abs(root->val - right.second));
            currMin = min(currMin, right.first);
            currMax = max(currMax, right.second);
        }

        return make_pair(currMin, currMax);
    }
};

int main()
{
    auto *solution = new Solution();

    auto* node4 = new TreeNode(4);
    auto* node7 = new TreeNode(7);
    auto* node1 = new TreeNode(1);
    auto* node13 = new TreeNode(13);
    auto* node6 = new TreeNode(6, node4, node7);
    auto* node14 = new TreeNode(14, node13, nullptr);
    auto* node3 = new TreeNode(3, node1, node6);
    auto* node10 = new TreeNode(10, nullptr, node14);
    auto* node8 = new TreeNode(8, node3, node10);
    cout << solution->maxAncestorDiff(node8) << endl;

    auto* node03 = new TreeNode(3);
    auto* node00 = new TreeNode(0, node03, nullptr);
    auto* node02 = new TreeNode(2, nullptr, node00);
    auto* node01 = new TreeNode(1, nullptr, node02);
    cout << solution->maxAncestorDiff(node01) << endl;

    delete solution;
    return 0;
}

/*
 Main idea:
 Зробити обхід дерева, на кожному кроці отримуємо мінімальне
 та максимальне значення правого та лівого піддерева. Перевіряємо
 чи поточний максимум є менший/більший

 Складність O(N): postorder
*/
